### PR TITLE
Added support for CNAME records add/remove

### PR DESCRIPTION
**Pull request by marank**
Signed-off-by: Matthias rank <development@m-rank.de>

**By submitting this pull request, I confirm the following:** `{please fill any appropriate checkboxes, e.g: [X]}`

`{Please ensure that your pull request is for the 'devel' branch!}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
- [X] I have Signed Off all commits. ([`git commit --signoff`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff))

---

**What does this PR aim to accomplish?:**
`{A detailed description, screenshots (if necessary), as well as links to any relevant GitHub issues}`
This PR enables adding/removing CNAME records via the admin interface.
To do so, the following PR needs to be accepted as well, because it allows editing the neccessary file using the pihole CLI: pi-hole/pi-hole#3351

**How does this PR accomplish the above?:**
`{A detailed description (such as a changelog) and screenshots (if necessary) of the implemented fix}`
grafik
I changed the menu accordingly to have both Local DNS records and CNAME records in one submenu.

**What documentation changes (if any) are needed to support this PR?:**
`{A detailed list of any necessary changes}`

### CODE REVIEW

**Comment by PromoFaux** [comment]
Hi there, thanks for the submission, I will try and find some time to review later on today

**Comment by PromoFaux** [comment]
Hi there,

OK, pulled a copy of this one and the accompanying core PR however I've fallen over at the first hurdle:

image

thus far clearing the cache has had no effect...

**Comment by marank** [comment]
That's what I get for accidentally testing my changes on another branch... Should work now, sorry for that!
On a side note, should I move the functions for managing the CNAME records also to func.php? Personally I think it's cleaner to keep them separated.

**Review by DL6ER — DISMISSED** [review]
I was thinking for a moment we should maybe fuse the DNS records and the CNAME records pages into one, however, it's likely clearer structured (esp. for the inexperienced) like you proposed it.

Some more thoughts:

- I merged the core PR into `development` already.
- There is a small merge conflict on this branch.

Yes, please move the routines over to `func.php` as you suggested. Just like I did in https://github.com/pi-hole/AdminLTE/pull/1276 and add support for importing CNAMEs through teleporter. There should be anything you need in #1276 where I did this for the local DNS records page.

Also, please add authentication checking for the CNAME page to avoid anyone being able to submit configuration even without logging into the web interface. See https://github.com/pi-hole/AdminLTE/commit/7b2d396295b41022ab9ed57ed00bca973b3c02ff for reference.

**Comment by marank** [comment]
I'll update this PR tomorrow. Thanks for your review!

**Comment by marank** [comment]
Okay, there were some hurdles while implementing the requested changes:

- I had to remove the definition of the `error` function inside of `AddCustomDNSEntry` because subsequent calls from teleporter failed (Cannot redeclare error() (previously declared in /var/www/html/admin/scripts/pi-hole/php/func.php:126) in /var/www/html/admin/scripts/pi-hole/php/func.php on line 126
- I had to move the declaration of the `$customDNSFile`/`$customCNAMEFile` vars to func.php because they didn't exist when called from teleporter
- When importing Local DNS Entries or Local CNAME Entries with teleporter and Clear existing data is set, every entry gets deleted seperately, always causing a restart of FTL (see webpage.sh). This has to be changed because FTL won't start if there were too many start attempts:
```
May 21 14:38:44 gradius systemd[1]: pihole-FTL.service: Start request repeated too quickly.
May 21 14:38:44 gradius systemd[1]: pihole-FTL.service: Failed with result 'start-limit-hit'.
```

I haven't used teleporter in the past, but all the tests I did were successfull.
On a side note, I'm not very familiar with resolving merge conflicts using git. I'd appreciate to get some help with that.

**Comment by DL6ER** [comment]
@marank Sorry for the delay on this. You can use the internal conflict resolution mechanism of Github. You should see it above the tests in this PR:


Screenshot at 2020-06-14 20-43-24

It will open a page showing what was added to the modified pages meanwhile so you can review the changes and decide what is the best when fusing your new changes with the ones made meanwhile, for instance:
Screenshot at 2020-06-14 20-45-53

**Comment by marank** [comment]
Ok, I'll need some time to update this PR. There were some changes in the devel branch in the meantime which I want to add.

**Comment by marank** [comment]
I think I'm done, looks good to me. The only issue remaining is the one with FTL restarting too often when restoring a backup via teleporter. Maybe we should add a flag to the pihole command which prevents restarting FTL after every relevant add/remove DNS/CNAME command?

**Review by DL6ER — APPROVED** [review]

**Comment by PromoFaux** [comment]
@DL6ER shall we change this to point at 5.1, or leave it til vNext?

**Comment by DL6ER** [comment]
Merge now for the next version after v5.1 gives us more time to test things properly.

File: scripts/pi-hole/js/customcname.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ *  This file is copyright under the latest version of the EUPL.
+ *  Please see LICENSE file for your rights under this license. */
+
+/* global utils:false */
+
+var table;
+var token = $("#token").text();
+
+function showAlert(type, message) {
```
Comment (XhmikosR) [inline scripts/pi-hole/js/customcname.js:13]: Why duplicate stuff here?

File: scripts/pi-hole/js/customcname.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
+/* Pi-hole: A black hole for Internet advertisements
+ *  (c) 2017 Pi-hole, LLC (https://pi-hole.net)
+ *  Network-wide ad blocking via your own hardware.
+ *
+ *  This file is copyright under the latest version of the EUPL.
+ *  Please see LICENSE file for your rights under this license. */
+
+/* global utils:false */
+
+var table;
```
Comment (XhmikosR) [inline scripts/pi-hole/js/customcname.js:10]: Please avoid global variables

**Comment by mcnahum** [comment]
I'm really waiting for that change to land.. any idea for when it is?

**Comment by dschaper** [comment]
When it's ready. Which is the same as "soon". 

Addressing the issues that were brought up in https://github.com/pi-hole/AdminLTE/pull/1278#issuecomment-655138561 would help.

**Pull request merged**